repo/fsrepo/datastores.go

```diff
@@ -0,0 +1,243 @@
+	flatfs "gx/ipfs/QmXZEfbEv9sXG9JnLoMNhREDMDgkq5Jd7uWJ7d77VJ4pxn/go-ds-flatfs"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	mount "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/syncmount"
+
+	levelds "gx/ipfs/QmaHHmfEozrrotyhyN44omJouyuEtx6ahddqV6W5yRaUSQ/go-ds-leveldb"
+	ldbopts "gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb/opt"
+)
+
+// ConfigFromMap creates a new datastore config from a map
+type ConfigFromMap func(map[string]interface{}) (DatastoreConfig, error)
+
+type DatastoreConfig interface {
+	// DiskId is a unique id representing the Datastore config as stored on disk, runtime config values are not
+	// part of this Id.  No length limit.
+	//DiskId() string
+
+	// Inst, Instantiate a new datastore from this config
+	Inst(path string) (repo.Datastore, error)
+}
+
+var datastores map[string]ConfigFromMap
+
+func init() {
+	datastores = map[string]ConfigFromMap{
+		"mount":   MountDatastoreConfig,
+		"flatfs":  FlatfsDatastoreConfig,
+		"levelds": LeveldsDatastoreConfig,
+		"mem":     MemDatastoreConfig,
+		"log":     LogDatastoreConfig,
+		"measure": MeasureDatastoreConfig,
+	}
+}
+
+func AnyDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	which, ok := params["type"].(string)
+	if !ok {
+		return nil, fmt.Errorf("'type' field missing or not a string")
+	}
+	fun, ok := datastores[which]
+	if !ok {
+		return nil, fmt.Errorf("unknown datastore type: %s", which)
+	}
+	return fun(params)
+}
+
+type mountDatastoreConfig struct {
+	mounts []premount
+}
+
+type premount struct {
+	ds     DatastoreConfig
+	prefix ds.Key
+}
+
+func MountDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	res := mountDatastoreConfig{}
+	mounts, ok := params["mounts"].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("'mounts' field is missing or not an array")
+	}
+	for _, iface := range mounts {
+		cfg := iface.(map[string]interface{})
+
+		child, err := AnyDatastoreConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		prefix, found := cfg["mountpoint"]
+		if !found {
+			return nil, fmt.Errorf("no 'mountpoint' on mount")
+		}
+
+		res.mounts = append(res.mounts, premount{
+			ds:     child,
+			prefix: ds.NewKey(prefix.(string)),
+		})
+	}
+
+	return res, nil
+}
+
+func (c mountDatastoreConfig) Inst(path string) (repo.Datastore, error) {
+	mounts := make([]mount.Mount, len(c.mounts))
+	for i, m := range c.mounts {
+		ds, err := m.ds.Inst(path)
+		if err != nil {
+			return nil, err
+		}
+		mounts[i].Datastore = ds
+		mounts[i].Prefix = m.prefix
+	}
+	return mount.New(mounts), nil
+}
+
+type flatfsDatastoreConfig struct {
+	cfg map[string]interface{}
+}
+
+func FlatfsDatastoreConfig(params map[string]interface{}) (DatastoreConfig, error) {
+	return flatfsDatastoreConfig{params}, nil
+}
+
+func (c flatfsDatastoreConfig) Inst(path string) (repo.Datastore, error) {
+	params := c.cfg
+	p, ok := params["path"].(string)
```

Comment:
maybe all this extraction of info from the generic `map[string]interface{}` should go in the config constructor above? would make the constructor here much cleaner